File: core/src/main/java/org/neo4j/importer/v1/targets/RelationshipTypeConstraint.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.neo4j.importer.v1.targets;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Objects;

public class RelationshipTypeConstraint extends Constraint {

    private final String property;

    @JsonCreator
    public RelationshipTypeConstraint(
            @JsonProperty(value = "name", required = true) String name,
            @JsonProperty(value = "property", required = true) String property) {

        super(name);
        this.property = property;
    }

    public String getProperty() {
        return property;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        RelationshipTypeConstraint that = (RelationshipTypeConstraint) o;
        return Objects.equals(property, that.property);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), property);
    }

    @Override
    public String toString() {
        return "RelationshipTypeConstraint{" + "property=" + property + ", options=" + super.toString();
    }
}

File: src/commonMain/kotlin/validate/relationship/constraint/RelationshipTypeConstraint.kt
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package validate.relationship.constraint

import model.GraphModel
import model.Relationship
import model.constraint.ConstraintType
import model.constraint.RelationshipConstraint
import validate.Issue
import validate.node.NodeConstraints
import validate.relationship.RelationshipValidation

object RelationshipTypeConstraint : RelationshipValidation {
    override fun dependsOn() = listOf(NodeConstraints)

    override fun validateConstraint(
        model: GraphModel,
        relationshipId: String,
        relationship: Relationship,
        constraintId: String,
        constraint: RelationshipConstraint,
        issues: MutableList<Issue>
    ) {
        if (constraint.type != ConstraintType.TYPE.name) {
            return
        }
        if (constraint.properties.size == 1) {
            return
        }
        issues.add(
            Issue(
                code = "invalid_relation_type_constraint_property_count",
                message = "Relationship type constraint '$constraintId' must have exactly one property.",
                path = "relationships.$relationshipId.constraints.$constraintId.properties"
            )
        )
    }
}
